### PR TITLE
feat(connector): support parsing "string containing JSON" as a struct type

### LINKED ISSUE #10772: connector: support parsing "string containing JSON"
Sometimes users may provide us JSONs that are not that normalized. For example, a JSON object can be first serialized into a string, then nested as a field of the parent object:

```json
{
  "nested": "{\"v1\": 10, \"v2\": \"hello\"}"
}
```

But users actually want it to be:
```json
{
  "nested": {
    "v1": 10,
    "v2": "hello"
  }
}
```

Currently, they have to specify the data type for `nested` as `varchar` in the source schema and then manually convert it into `jsonb` as a workaround. However, manipulating the `JsonbArray` can be inefficient, especially for nested `jsonb` since we currently have to allocate a new `JsonbArray` for the intermediate accessing result.

It'll be better if we can be more flexible on parsing these "strings containing JSON" so that users can specify these nested fields seamlessly just as usual.

There's a [`serde_with::json::JsonString`](https://docs.rs/serde_with/latest/serde_with/json/struct.JsonString.html) which has a similar idea to this.

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As explained in #10772. Now we can parse
```json
{
  "nested": "{\"v1\": 10, \"v2\": \"hello\"}"
}
```
as `CREATE SOURCE .. (nested struct<v1 int, v2 varchar>) FORMAT JSON`.

Close #10772.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Support parsing "string containing JSON" as a struct type.

### CODE REVIEW

File: src/connector/src/parser/unified/json.rs [hunk 1]
```diff
@@ -446,6 +446,17 @@ impl JsonParseOptions {
...
+            // String containing json object, e.g. "{\"a\": 1, \"b\": 2}"
+            // Try to parse it as json object.
+            (Some(DataType::Struct(_)), ValueType::String) => {
+                // TODO: avoid copy by accepting `&mut BorrowedValue` in `parse` method.
+                let mut value = value.as_str().unwrap().as_bytes().to_vec();
+                let value =
+                    simd_json::to_borrowed_value(&mut value[..]).map_err(|_| create_error())?;
+                return self.parse(&value, type_expected);
+            }
+
```
Comment (tabVersion) [inline src/connector/src/parser/unified/json.rs:459]: In the description, the schema should be declared as `nested struct <>` but there is no option for `nested`

**Review by tabVersion — APPROVED** [review]
rest LGTM, please consider making it an parser option like `JsonParseOptions`

File: src/connector/src/parser/unified/json.rs (hunk 1 above)
Comment (BugenZhao) [reply src/connector/src/parser/unified/json.rs:459]: Ahh, `nested` is the name of the column as an example. 😂

**Review by xiangjinwu — APPROVED** [review]
lgtm

**Comment by BugenZhao** [comment]
> please consider making it an parser option like `JsonParseOptions`

Sure, while we seem not to provide a way to customize the options by users. Do you think we should by default enable this?

**Comment by BugenZhao** [comment]
> Do you think we should by default enable this?

Well, I've enabled it for `DEFAULT` options and left Canal and Debezium `Strict`, as they are generated by the connector and tend to be more normalized.

**Comment by tabVersion** [comment]
> > please consider making it an parser option like `JsonParseOptions`
> 
> Sure, while we seem not to provide a way to customize the options by users. Do you think we should by default enable this?

sorry I missed the message. it will be available in parser options in the future.

**Pull request merged**